File: ZMQ/lec-05-prg-09-dealer-router-async-server.py
# Asynchronous Client/Server Pattern
# Reference: https://zguide.zeromq.org/docs/chapter3/#The-Asynchronous-Client-Server-Pattern
# author of original code: Felipe Cruz <felipecruz@loogica.net>
# license of original code: MIT/X11

import zmq
import sys
import threading
import time
from random import randint, random

class ServerTask(threading.Thread):
    """ServerTask"""
    def __init__(self, num_server):
        threading.Thread.__init__ (self)
        self.num_server = num_server

    def run(self):
        context = zmq.Context()
        frontend = context.socket(zmq.ROUTER)
        frontend.bind('tcp://*:5570')

        backend = context.socket(zmq.DEALER)
        backend.bind('inproc://backend')

        workers = []
        for i in range(self.num_server):
            worker = ServerWorker(context, i)
            worker.start()
            workers.append(worker)

        zmq.proxy(frontend, backend) 

        frontend.close()
        backend.close()
        context.term()

class ServerWorker(threading.Thread):
    """ServerWorker"""
    def __init__(self, context, id):
        threading.Thread.__init__ (self)
        self.context = context
        self.id = id

    def run(self):
        worker = self.context.socket(zmq.DEALER)
        worker.connect('inproc://backend')
        print('Worker#{0} started'.format(self.id))
        while True:
            ident, msg = worker.recv_multipart()
            print('Worker#{0} received {1} from {2}'.format(self.id, msg, ident))
            worker.send_multipart([ident, msg])

        worker.close() # useless

def main(argv):
    """main function"""
    server = ServerTask(int(argv[1]))
    server.start()
    server.join()

# usage: python 07-dealer-router-async-server.py n
if __name__ == "__main__":
    main(sys.argv)

File: ZMQ/lec-05-prg-09-dealer-router-async-server.js
const zmq = require("zeromq");

class ServerTask {
  constructor(num_server) {
    this.num_server = num_server;
  }

  async run() {
    const frontend = new zmq.Router();
    await frontend.bind("tcp://*:5570");

    const backend = new zmq.Dealer();
    await backend.bind("inproc://backend");

    const workers = [];

    for (let i = 0; i < this.num_server; i++) {
      const worker = new ServerWorker(i);
      worker.run();
      workers.push(worker);
    }

    const proxy = new zmq.Proxy(frontend, backend);
    proxy.run();
  }
}

class ServerWorker {
  constructor(id) {
    this.id = id;
  }

  async run() {
    const worker = new zmq.Dealer();
    worker.connect("inproc://backend");
    console.log(`Worker#${this.id} started`);

    for await (const [ident, msg] of worker) {
      console.log(`Worker#${this.id} received ${msg} from ${ident}`);
      await worker.send([ident, ident, msg]);
    }
  }
}

const server = new ServerTask(Number(process.argv[2]));
server.run();
